File: templates/api-client.ts
/**
 * TypeScript API Client: Retry, timeout, typed errors
 * Usage: const { data, error } = await typedFetch<User[]>("/api/users");
 */

interface ApiResponse<T> {
  data: T | null;
  error: { code: string; message: string; status: number } | null;
}

interface ApiCallOptions {
  retries?: number;
  timeoutMs?: number;
}

export async function apiCall<T>(
  fn: (signal: AbortSignal) => Promise<T>,
  options?: ApiCallOptions
): Promise<ApiResponse<T>> {
  const { retries = 3, timeoutMs = 10000 } = options ?? {};

  for (let attempt = 0; attempt <= retries; attempt++) {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), timeoutMs);
    try {
      const result = await fn(controller.signal);
      clearTimeout(timeout);
      return { data: result, error: null };
    } catch (err) {
      clearTimeout(timeout);
      if (attempt === retries) {
        const isTimeout = err instanceof DOMException && err.name === "AbortError";
        return {
          data: null,
          error: {
            code: isTimeout ? "TIMEOUT" : "API_ERROR",
            message: err instanceof Error ? err.message : "Unknown error",
            status: isTimeout ? 408 : 500,
          },
        };
      }
      await new Promise((r) => setTimeout(r, 1000 * 2 ** attempt));
    }
  }
  return { data: null, error: { code: "UNREACHABLE", message: "Max retries", status: 500 } };
}

export async function typedFetch<T>(
  path: string,
  options?: RequestInit & { baseUrl?: string; retries?: number; timeoutMs?: number }
): Promise<ApiResponse<T>> {
  const { baseUrl = "", retries = 3, timeoutMs = 10000, ...fetchOptions } = options ?? {};
  return apiCall<T>(async (signal) => {
    const response = await fetch(`${baseUrl}${path}`, {
      ...fetchOptions,
      signal,
      headers: { "Content-Type": "application/json", ...fetchOptions.headers },
    });
    if (!response.ok) throw new Error(`${response.status}: ${await response.text()}`);
    return response.json();
  }, { retries, timeoutMs });
}

File: templates/api_client.py
"""
Python Async API Client — httpx + tenacity retry
Usage:
    async with ExternalApiClient("https://api.example.com") as api:
        result = await api.get("/users")
        if result.ok: print(result.data)
"""
from __future__ import annotations
import httpx
from pydantic import BaseModel
from tenacity import retry, stop_after_attempt, wait_exponential, retry_if_exception_type
from typing import Any


class ApiResponse(BaseModel):
    data: Any | None = None
    error: str | None = None
    status: int = 200

    @property
    def ok(self) -> bool:
        return self.error is None and 200 <= self.status < 300


class ExternalApiClient:
    def __init__(self, base_url: str, timeout: float = 10.0, headers: dict[str, str] | None = None):
        self.client = httpx.AsyncClient(base_url=base_url, timeout=timeout, headers=headers or {})

    async def __aenter__(self): return self
    async def __aexit__(self, *args): await self.close()

    @retry(stop=stop_after_attempt(3), wait=wait_exponential(multiplier=1, min=1, max=10),
           retry=retry_if_exception_type((httpx.TimeoutException, httpx.ConnectError)))
    async def _request(self, method: str, path: str, **kwargs) -> httpx.Response:
        response = await self.client.request(method, path, **kwargs)
        response.raise_for_status()
        return response

    async def _safe(self, method: str, path: str, **kwargs) -> ApiResponse:
        try:
            resp = await self._request(method, path, **kwargs)
            return ApiResponse(data=resp.json(), status=resp.status_code)
        except httpx.HTTPStatusError as e:
            return ApiResponse(error=str(e), status=e.response.status_code)
        except Exception as e:
            return ApiResponse(error=str(e), status=500)

    async def get(self, path: str, **kw) -> ApiResponse: return await self._safe("GET", path, **kw)
    async def post(self, path: str, **kw) -> ApiResponse: return await self._safe("POST", path, **kw)
    async def put(self, path: str, **kw) -> ApiResponse: return await self._safe("PUT", path, **kw)
    async def delete(self, path: str, **kw) -> ApiResponse: return await self._safe("DELETE", path, **kw)
    async def close(self): await self.client.aclose()
